### ISSUE TITLE
Replies to messages sent from non-Matrix clients get dropped until restart in a single-puppeted bridge

### ISSUE THREAD

**Issue by cincodenada** [mautrix/facebook#173]
## Reproduction

1. Set up a single-puppeted Mautrix Facebook bridge
2. Send a message to someone using the Facebook app
3. Have them reply to that message.

## Expected behavior

The message shows up immediately, optimally with the reply attached

## Actual behavior

An error is logged, the message doesn't show up until the bridge is restarted, and even then it doesn't show the message being replied to, as it does if the message was sent from Matrix. Here's what the reproduction looks like from Element and Facebook - you can't tell from a screenshot of course, but the "This is a reply from my fake friend" message doesn't show up in Matrix until after a restart of the bridge. Also, I didn't cut anything off, my "This is a message I am sending to my fake friend" message doesn't show up in Matrix at all, which is expected given that I don't have double-puppeting set up.
![fakemessagetwo](https://user-images.githubusercontent.com/479715/130374983-7209df04-3d33-452c-9b83-46bbdf46d060.png)
![MatrixVersion](https://user-images.githubusercontent.com/479715/130374988-877034fa-f27d-4cb2-a586-c0033b358829.png)

## Details

When my friend replied to my message, the bridge couldn't find the replied-to message, threw an error, and dropped my friend's reply. That's what this error looked like:
```
[2021-08-22 16:39:20,163] [DEBUG@mau.portal.100068548189650<->42904324] Handling Facebook event mid.$cAABbAwTKNtaBoYDkw17cDucMVabw (/6835354719948744432)
[2021-08-22 16:39:20,164] [WARNING@mau.portal.100068548189650<->42904324] Ignoring own message mid.$cAABbAwTKNtaBoYDkw17cDucMVabw in private chat because own puppet is not in room.
[2021-08-22 16:39:22,571] [DEBUG@mau.portal.100068548189650<->42904324] Handled Messenger read receipt from 100068548189650 up to 1629675562196/$_QGas13jWAow2yGcg8sUt46dOKy2gDWCBuoI_6FMpbI
[2021-08-22 16:39:34,006] [DEBUG@mau.portal.100068548189650<->42904324] Handling Facebook event mid.$cAABbAwTKNtaBoYEasl7cDvSWB3zs (/6835354778093452524)
[2021-08-22 16:39:34,007] [ERROR@mau.portal.100068548189650<->42904324] Error handling Facebook message mid.$cAABbAwTKNtaBoYEasl7cDvSWB3zs
Traceback (most recent call last):
  File "/home/joel/mautrix/mautrix-facebook/lib/python3.8/site-packages/mautrix_facebook/portal.py", line 771, in handle_facebook_message
    await self._handle_facebook_message(source, sender, message, reply_to)
  File "/home/joel/mautrix/mautrix-facebook/lib/python3.8/site-packages/mautrix_facebook/portal.py", line 840, in _handle_facebook_message
    event_ids = await self._handle_mqtt_message(source, intent, message, reply_to)
  File "/home/joel/mautrix/mautrix-facebook/lib/python3.8/site-packages/mautrix_facebook/portal.py", line 868, in _handle_mqtt_message
    event_ids.append(await self._handle_facebook_text(intent, message, reply_to,
  File "/home/joel/mautrix/mautrix-facebook/lib/python3.8/site-packages/mautrix_facebook/portal.py", line 1018, in _handle_facebook_text
    await self._add_facebook_reply(content, reply_to)
  File "/home/joel/mautrix/mautrix-facebook/lib/python3.8/site-packages/mautrix_facebook/portal.py", line 731, in _add_facebook_reply
    if not message.fbid:
AttributeError: 'NoneType' object has no attribute 'fbid'
[2021-08-22 16:39:34,054] [DEBUG@mau.portal.100068548189650<->42904324] Handled Messenger read receipt from 100068548189650 up to 1629675573848/$_QGas13jWAow2yGcg8sUt46dOKy2gDWCBuoI_6FMpbI
```
That error took me to [this line](https://github.com/mautrix/facebook/blob/master/mautrix_facebook/portal.py#L767) in `portal.py` - `message` there was empty, as expected given that the message wasn't in the database. On restart it seemed to take a different code path when processing that message (parsing it as a `MinimalMessage`, maybe?):
```
[2021-08-22 16:41:12,536] [DEBUG@mau.portal.100068548189650<->42904324] Got 2 messages from server
[2021-08-22 16:41:12,542] [DEBUG@mau.portal.100068548189650<->42904324] Handling Facebook event mid.$cAABbAwTKNtaBoYDkw17cDucMVabw (/6835354719948744432)
[2021-08-22 16:41:12,543] [WARNING@mau.portal.100068548189650<->42904324] Ignoring own message mid.$cAABbAwTKNtaBoYDkw17cDucMVabw in private chat because own puppet is not in room.
[2021-08-22 16:41:12,544] [DEBUG@mau.portal.100068548189650<->42904324] Handling Facebook event mid.$cAABbAwTKNtaBoYEasl7cDvSWB3zs (/6835354778093452524)
[2021-08-22 16:41:12,545] [WARNING@mau.portal.100068548189650<->42904324] Couldn't find reply target MinimalMessage(id='bWVzc2FnZTptaWQuJGNBQUJiQXdUS050YUJvWURrdzE3Y0R1Y01WYWJ3', message_id='mid.$cAABbAwTKNtaBoYDkw17cDucMVabw', message=MessageText(text='This is a message I am sending to my fake friend', ranges=[]), message_sender=MessageSender(id='42904324', messaging_actor=MinimalParticipant(id='42904324', name=None)), sticker=None, blob_attachments=[], extensible_attachment=None) to bridge reply metadata to Matrix
[2021-08-22 16:41:12,611] [DEBUG@mau.portal.100068548189650<->42904324] Handled Messenger message mid.$cAABbAwTKNtaBoYEasl7cDvSWB3zs -> ['$9F-w1YHz2M00-s1n3YqbiHAFTWhPTJp211sdwz9xa5c']
[2021-08-22 16:41:12,626] [INFO@mau.portal.100068548189650<->42904324] Backfilled 2 messages through @joel:c9a.dev
```
This time didn't error out, so after the restart shown above, my friend's message _did_ make it through to Matrix, but of course without its message, as shown above.

## Workaround

Obviously restarting is a workaround, but that's not great. I changed the code that caused the error as follows:
```
            if not message:
                pass # Fall through to warning below
            elif not message.fbid:
                self.log.debug(f"Got message ID {meta.id} for offline threading ID "
                               f"{message.fb_txn_id} / {message.mxid} (in database) from reply")
                message.fbid = meta.id
                message.timestamp = meta.timestamp
                await message.update()
```
And with that change, the message showed up as expected, albeit still without the message being replied to.

I tried digging in a little further, because as shown in the `WARNING` line at `2021-08-22 16:41:12,545` in the last error log, `reply_to` in this function seems to have at least the text being replied to already in the object in the `graphql.MinimalMessage` case, and I also saw it in the `mqtt.Message` case, so I hoped I could just set the reply-to text from that, but it seems like `set_reply()` expects either an id (that I assume then tries to look it up or something, which will again fail?) or a fuller `MessageEvent` which requires room_id and things that I didn't know where to get if I were to construct one from scratch using the message in `reply_to`.

**Comment by Nico264** [comment]
Even without double puppeting, you can have your own messages sent from other clients bridged to Matrix by setting `bridge.invite_own_puppet_to_pm` to `true`. (let's call that 1.5-puppeting)
This is the direct answer for the first warning in your log: `[2021-08-22 16:41:12,543] [WARNING@mau.portal.100068548189650<->42904324] Ignoring own message mid.$cAABbAwTKNtaBoYDkw17cDucMVabw in private chat because own puppet is not in room.`.

I believe this would also workaround the reply issue (untested).

**Comment by r1bnc** [comment]
Any update to this? I seem to get an error as well but I don't receive the warning: ignoring own message instead it just displays:

```
 Error handling Facebook message mid...
...
Type' object has no attribute 'fbid'
```
Using 0.3.1 stable

**Comment by cincodenada** [comment]
@r1bnc Can you update your paste to include the rest of the message that is currently omitted? There's no sensitive information in there (although you could remove the message ID to be extra sure) and that would help confirm if it's the same issue or not.

**Comment by r1bnc** [comment]
```
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]: [2021-09-24 02:56:48,905] [DEBUG@mau.portal.5XXXXXXX] Handled Messenger message mid.$XXXXX -> ['$XXX-XXXX-XXX']
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]: [2021-09-24 02:56:48,916] [DEBUG@mau.portal.1XXXXX1] Handling Facebook event mid.$XXX-XXXX (/684XX0)
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]: [2021-09-24 02:56:48,920] [ERROR@mau.portal.1XXXXX1] Error handling Facebook message  mid.$XXX-XXXX
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]: Traceback (most recent call last):
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]:   File "/usr/lib/python3.9/site-packages/mautrix_facebook/portal.py", line 774, in handle_facebook_message
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]:     await self._handle_facebook_message(source, sender, message, reply_to)
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]:   File "/usr/lib/python3.9/site-packages/mautrix_facebook/portal.py", line 843, in _handle_facebook_message
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]:     event_ids = await self._handle_mqtt_message(source, intent, message, reply_to)
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]:   File "/usr/lib/python3.9/site-packages/mautrix_facebook/portal.py", line 871, in _handle_mqtt_message
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]:     event_ids.append(await self._handle_facebook_text(intent, message, reply_to,
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]:   File "/usr/lib/python3.9/site-packages/mautrix_facebook/portal.py", line 1021, in _handle_facebook_text
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]:     await self._add_facebook_reply(content, reply_to)
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]:   File "/usr/lib/python3.9/site-packages/mautrix_facebook/portal.py", line 734, in _add_facebook_reply
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]:     if not message.fbid:
Sep 24 10:56:48 localhost.localdomain matrix-mautrix-facebook[3697250]: AttributeError: 'NoneType' object has no attribute 'fbid'

```

**Comment by cincodenada** [comment]
Thanks, that does look like the same error, although it could have a different root cause. Did you try setting `bridge.invite_own_puppet_to_pm` to `true` in your `config.yaml` for `mautrix-facebook`? I just did that and restarted the bridge, and it seems to have worked great - messages I send from Facebook's clients show up in Matrix, and it does also seem to be working around the message-reply issue.

I'll leave this issue open, and perhaps open a PR with my suggested fix above, since I think even without 1.5-puppeting set up as above the bridge should be showing the replies, even if it drops the text being replied to.

**Comment by r1bnc** [comment]
I'll try that suggestion, thank you.
